src/sources/youtube.rs

```diff
@@ -0,0 +1,170 @@
+use crate::sources::ffmpeg::ffmpeg_yt;
+use serde_json::Value;
+use serenity::async_trait;
+use songbird::input::{
+    error::{Error as SongbirdError, Result as SongbirdResult},
+    restartable::Restart,
+    Codec, Container, Input, Metadata, Restartable,
+};
+use std::{
+    io::{BufRead, BufReader, Read},
+    process::Command,
+    process::{Child, Stdio},
+    time::Duration,
+};
+use tokio::task;
+
+pub struct YouTubeRestartable {}
+
+impl YouTubeRestartable {
+    pub async fn ytdl<P: AsRef<str> + Send + Clone + Sync + 'static>(
+        uri: P,
+        lazy: bool,
+    ) -> SongbirdResult<Restartable> {
+        Restartable::new(YouTubeRestarter { uri }, lazy).await
+    }
+
+    pub async fn ytdl_search<P: AsRef<str> + Send + Clone + Sync + 'static>(
+        uri: P,
+        lazy: bool,
+    ) -> SongbirdResult<Restartable> {
+        let uri = format!("ytsearch1:{}", uri.as_ref());
+        Restartable::new(YouTubeRestarter { uri }, lazy).await
+    }
+
+    pub async fn ytdl_playlist(uri: &str) -> Option<Vec<String>> {
+        let mut child = Command::new("yt-dlp")
+            .args([uri, "--flat-playlist", "-j"])
+            .stdout(Stdio::piped())
+            .spawn()
+            .unwrap();
+
+        if let Some(stdout) = &mut child.stdout {
+            let reader = BufReader::new(stdout);
+
+            let lines = reader.lines().flatten().map(|line| {
+                let entry: Value = serde_json::from_str(&line).unwrap();
+                entry.get("url").unwrap().as_str().unwrap().to_string()
+            });
+
+            Some(lines.collect())
+        } else {
+            None
+        }
+    }
+}
+
+struct YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Sync,
+{
+    uri: P,
+}
+
+#[async_trait]
+impl<P> Restart for YouTubeRestarter<P>
+where
+    P: AsRef<str> + Send + Clone + Sync,
+{
+    async fn call_restart(&mut self, time: Option<Duration>) -> SongbirdResult<Input> {
+        let (yt, metadata) = ytdl(self.uri.as_ref()).await?;
+
+        if let Some(time) = time {
+            let ts = format!("{:.3}", time.as_secs_f64());
+            ffmpeg_yt(yt, metadata, &["-ss", &ts]).await
+        } else {
+            ffmpeg_yt(yt, metadata, &[]).await
+        }
+    }
+
+    async fn lazy_init(&mut self) -> SongbirdResult<(Option<Metadata>, Codec, Container)> {
+        _ytdl_metadata(self.uri.as_ref())
+            .await
+            .map(|m| (Some(m), Codec::FloatPcm, Container::Raw))
+    }
+}
+
+async fn ytdl(uri: &str) -> Result<(Child, Metadata), SongbirdError> {
+    let ytdl_args = [
+        "-j",            // print JSON information for video for metadata
+        "--no-simulate", // ensure video is downloaded regardless of printing
+        "-f",
+        "bestaudio", // select best quality audio-only
+        "-R",
+        "infinite",        // infinite number of download retries
```

Comment:
A possible problem in the future, a maximum number, even if high, should be used I would say

Reply:
Also, why is the comment so spaced out 🥴

Reply:
Ye, true, hence "in the future". The spacing is weird, fix that too @aquelemiguel .

Reply:
Not my fault, fmt does that. 🤷